storage/innobase/trx/trx0purge.cc

```diff
@@ -428,16 +443,24 @@ trx_purge_truncate_rseg_history(trx_rseg_t &rseg,
   const trx_id_t undo_trx_no=
     mach_read_from_8(b->page.frame + hdr_addr.boffset + TRX_UNDO_TRX_NO);
 
-  if (undo_trx_no >= limit.trx_no)
+  if (undo_trx_no >= trx_no)
   {
-    if (undo_trx_no == limit.trx_no)
-      err = trx_undo_truncate_start(&rseg, hdr_addr.page,
-                                    hdr_addr.boffset, limit.undo_no);
+    if (undo_trx_no == trx_no)
+      err= trx_undo_truncate_start(&rseg, hdr_addr.page,
+                                   hdr_addr.boffset, undo_no);
     goto func_exit;
   }
-
-  if (!all)
-    goto func_exit;
+  else
+  {
+    rseg_ref= rseg.ref_load();
+    if (rseg_ref >= rseg.REF || !purge_sys.sees(rseg.needs_purge))
+    {
+      if (!(rseg.SKIP & rseg_ref) && !freed &&
+          &rseg == &trx_sys.rseg_array[purge_sys.m_skipped_rseg])
+        rseg.set_skip_allocation();
```

Comment:
This could increase the probability of [a debug assertion failure](https://buildbot.mariadb.org/#/builders/572/builds/5615/steps/8/logs/stdio):
```
CURRENT_TEST: innodb.innodb_bug12400341
mysqltest: At line 61: query 'UPDATE mysqltest.transtable SET val = 1 WHERE id = 33 - $c' failed: <Unknown> (2013): Lost connection to server during query
The result from queries just before the failure was:
call mtr.add_suppression("\\[Warning\\] InnoDB: Cannot find a free slot for an undo log. Do you have too");
drop database if exists mysqltest;
create database mysqltest;
CREATE TABLE mysqltest.transtable (id int unsigned NOT NULL PRIMARY KEY, val int DEFAULT 0) ENGINE=InnoDB;
select count(*) from information_schema.processlist where command != 'Daemon';
count(*)
33
Server [mysqld.1 - pid: 159238, winpid: 159238, exit: 256] failed during test run
Server log from this test:
----------SERVER LOG START-----------
mariadbd: /home/buildbot/amd64-debian-11-msan/build/storage/innobase/trx/trx0trx.cc:823: void trx_assign_rseg_low(trx_t *): Assertion `!look_for_rollover || start_scan_slot != slot' failed.
```
This test is playing with `innodb_trx_rseg_n_slots_debug`. I’ll try to reproduce the failure. I’d need a core dump to check if there were multiple skipped rollback segments at the time of the failure, or if `purge_sys.m_skipped_rseg` could have been changing during that traversal.

Reply:
I was able to reproduce this with a simple invocation:
```bash
./mtr --parallel=auto --repeat=10 innodb.innodb_bug12400341{,,,,,,,,,,,,,,,,}
```
Several workers had hung by the time most workers had completed the 10 passes. The failure occurred on the 7th test repetition of one server:
```
innodb.innodb_bug12400341 'innodb'       w4 [ 7 fail ]
        Test ended at 2024-01-12 09:25:21

CURRENT_TEST: innodb.innodb_bug12400341
mysqltest: At line 61: query 'UPDATE mysqltest.transtable SET val = 1 WHERE id = 33 - $c' failed: <Unknown> (2013): Lost connection to server during query
```
In the core dump, I see `slot=start_scan_slot=0` and `purge_sys.m_skipped_rseg=1`. All `trx_sys.rseg_array[].ref` are 0. It would seem to me that `purge_sys_t::iterator::free_history_rseg()` had just re-enabled the slot 0 and advanced `purge_sys.m_skipped_rseg`. When the debug variable `trx_rseg_n_slots_debug` is set, `trx_assign_rseg_low()` will use only `slot=0`. Thus, we’d better add a condition to disable the throttling when that debug variable is set.